### PR TITLE
Initial FPGA Support

### PR DESCRIPTION
**Pull request by TizianoDeMatteis**
This PR includes several ONNX Node expansions specifically tailored for FPGA, in particular IntelFPGA.

### CODE REVIEW

File: daceml/onnx/op_implementations/fpga_implementations.py [hunk 1]
```diff
@@ -0,0 +1,2600 @@
...
+                "im2col_map",
+                {
+                    "b": "0:{}".format(batch_size),
+                    "n": "0:{}/{}".format(
+                        num_filters, P),  # repeat B for computing the result
+                    "cin": "0:{}".format(num_channels),
+                    "hx": "0:{}".format(filter_hx),
+                    "hy": "0:{}".format(filter_hy),
+                    "x": "0:{}".format(output_size_x),
+                    "y0": "0:{}/{}".format(output_size_x, vec_width),
```
Comment (manuelburger) [inline daceml/onnx/op_implementations/fpga_implementations.py:543]:
```suggestion
                    "y0": "0:{}/{}".format(output_size_y),
```
I think `output_size_y` already accounts for vectorization; if `output_size_x` is used I think we should add a test to verify the input is square-shaped.

**Review by manuelburger — COMMENTED** [review]
Added a comment in FPGA Im2Col implementation

File: tests/pytorch/fpga/test_reshape_fpga.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    sdfg = dace_model.sdfg
+    sdfg.apply_transformations([FPGATransformSDFG])
+
+    donnx.ONNXReshape.default_implementation = 'fpga'
+    sdfg.expand_library_nodes()
+    sdfg.apply_transformations_repeated([InlineSDFG])
+
+    dace_output_fpga = dace_model(x)
+    dace_output_fpga = dace_output_fpga.reshape(
+        torch_output.detach().numpy().shape)
```
Comment (manuelburger) [inline tests/pytorch/fpga/test_reshape_fpga.py:55]:
```suggestion
        torch_output.detach().numpy().shape).detach().numpy()
```
Need to detach and convert the reshaped tensor to NumPy, to be able to use `np.linalg.norm` on the result below.

Comment (TizianoDeMatteis) [reply tests/pytorch/fpga/test_reshape_fpga.py:55]: Thanks!

File: .codecov.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+coverage:
```
Comment (orausch) [inline .codecov.yml:1]: Please remove this file

File: daceml/onnx/nodes/codegen.py [hunk 4]
```diff
@@ -327,7 +327,7 @@ def expand_node(node, state, sdfg):
     inputs_on_host = [True for _ in range(len(inputs))]
 
     actual_node_schedule = node.schedule
-    if node.schedule == dtypes.ScheduleType.CPU_Multicore or node.schedule == dtypes.ScheduleType.Default:
+    if node.schedule == dtypes.ScheduleType.CPU_Multicore or node.schedule == dtypes.ScheduleType.Default or node.schedule == dtypes.ScheduleType.Sequential:
```
Comment (orausch) [inline daceml/onnx/nodes/codegen.py:330]: Not sure if this is a good idea. What happens if an ORT expansion is called within a OpenMP parallel map?

File: daceml/onnx/op_implementations/fpga_implementations.py [hunk 5]
```diff
@@ -0,0 +1,3026 @@
...
+from daceml.util.utils import in_desc_with_name, out_desc_with_name, in_edge_with_name
+from daceml.transformation import constant_folding
+
+
+def _2d_sliding_window_index_expr(x_or_y, stride, kernel_size):
+    index_expression = "out_{x_or_y} * {stride} + h{x_or_y}"
+    return index_expression.format(x_or_y=x_or_y, stride=stride)
+
+
+def program_for_node(program, sdfg: SDFG, state: SDFGState,
```
Comment (orausch) [inline daceml/onnx/op_implementations/fpga_implementations.py:28]: Please import the version in the utils file.

File: daceml/onnx/op_implementations/fpga_implementations.py [hunk 6]
```diff
@@ -0,0 +1,3026 @@
...
+                    name, node.schema.name))
+
+    program.__annotations__ = annotations
+
+    result = DaceProgram(program, (), {}, False, 0)
+
+    return result
+
+
+@autoregister_params(op="Conv", name="naive_fpga")
```
Comment (orausch) [inline daceml/onnx/op_implementations/fpga_implementations.py:64]: We now have an `op_implementation` decorator, please use that.

File: daceml/util/utils.py [hunk 7]
```diff
@@ -18,6 +19,19 @@
 log = logging.getLogger(__name__)
 
 
+def is_desc_contiguous(desc: dt.Data) -> bool:
```
Comment (orausch) [inline daceml/util/utils.py:22]: Please remove this function here (it's already defined below)

File: tests/pytorch/test_attn.py [hunk 8]
```diff
@@ -9,7 +9,7 @@
 
 
 @pytest.mark.ort
-def test_attn(gpu, sdfg_name):
+def test_attn():
```
Comment (orausch) [inline tests/pytorch/test_attn.py:12]: Please undo changes in this file

**Review by orausch — CHANGES_REQUESTED** [review]
Quick look over the PR and some comments on high level things. I didn't review the tests or the FPGA operator implementations.

What's the status on testing? Can we add the testing script to the CI?

**Comment by TizianoDeMatteis** [comment]
> Can we add the testing script to the CI?

Similarly to what is done with Dace, we would need to set up an access to an external machine where we have installed all the FPGA software. @tbennun do you think this is possible?

File: daceml/onnx/nodes/codegen.py (hunk 4 above)
Comment (TizianoDeMatteis) [reply daceml/onnx/nodes/codegen.py:330]: Yep, that was the case

**Comment by orausch** [comment]
> we would need to set up an access to an external machine where we have installed all the FPGA software

The GPU tests run on the same machine as in Dace

**Comment by TizianoDeMatteis** [comment]
Then yes, definitely

File: daceml/onnx/shape_inference/symbolic_shape_infer.py [hunk 9]
```diff
@@ -21,19 +21,26 @@ def get_attribute(node, attr_name, default_value=None):
 
 
 def get_dim_from_type_proto(dim):
-    return getattr(dim, dim.WhichOneof('value')) if type(dim.WhichOneof('value')) == str else None
+    return getattr(dim, dim.WhichOneof('value')) if type(
```
Comment (orausch) [inline daceml/onnx/shape_inference/symbolic_shape_infer.py:24]: Please revert formatting changes in this file: we don't maintain it, it's just mirrored from ONNX Runtime

File: examples/lenet.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
+""" A lenet inference script. Example adapted from https://github.com/pytorch/examples/blob/master/mnist/main.py """
```
Comment (orausch) [inline examples/lenet.py:1]: Can we rename this to lenet_fpga or something? It's more of an FPGA example.

File: pytest.ini [hunk 11]
```diff
@@ -1,6 +1,8 @@
 [pytest]
+;addopts = --tb=short
```
Comment (orausch) [inline pytest.ini:2]: Please remove this

File: tests/pytorch/fpga/fpga_testing.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python3
```
Comment (orausch) [inline tests/pytorch/fpga/fpga_testing.py:1]: Is this file still required?

**Review by orausch — CHANGES_REQUESTED** [review]
Small comments, will look at the failure now

File: tests/pytorch/fpga/test_conv2d_fpga.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+from daceml.util import utils
+from dace.transformation.interstate import FPGATransformSDFG, InlineSDFG
+from daceml.transformation import InputToConstant
+from dace.transformation.dataflow import streaming_memory as sm
+from dace.transformation.dataflow import PruneConnectors
+from multiprocessing import Process, Queue
+
+import daceml.onnx as donnx
+
+donnx.default_implementation = "pure"
```
Comment (orausch) [inline tests/pytorch/fpga/test_conv2d_fpga.py:22]: Please don't overwrite the global default implementation like this.
We have the `@pytest.mark.pure` decorator for tests that will set the implementation to pure and reset it afterwards.

File: tests/pytorch/fpga/test_conv2d_fpga.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    dace_model = DaceModule(ptmodel, dummy_inputs=x, auto_optimize=False)
+
+    if execute_cpu_dace:
+        dace_output = dace_model(x)
+
+    sdfg = dace_model.sdfg
+
+    ###################################################
+    # Transform for FPGA and Inline
+    donnx.ONNXConv.default_implementation = "naive_fpga"
```
Comment (orausch) [inline tests/pytorch/fpga/test_conv2d_fpga.py:73]: You can use `with dace.library.change_default(donnx.ONNXConv, "naive_fpga"):`

**Review by orausch — CHANGES_REQUESTED** [review]
Took me a while to investigate, but the reason the tests fail only when everything is run and not when run individually is because some FPGA tests overwrite the default implementation on default.
Everywhere where you set `default_implementation`, you should make sure that it is reset afterwards. You can usually use `dace.library.change_default` and the pytest `@pytest.mark.pure` decorator for this.

**Comment by TizianoDeMatteis** [comment]
@orausch  I've fixed this, but the error seems to persist.
It is raised even if I explicitly ignore the fpga test directory with pytest.

How did you test it locally?

**Comment by orausch** [comment]
Atleast two of the errors are due to a strict transform error that is also currently broken on master. I believe Tal is working on it

File: daceml/transformation/input_to_constant.py [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+            Tensor) else parent.weights[unclean_onnx_name]
+        sdfg.add_constant(data_name, data, sdfg.arrays[node.data])
+
+        for out_edge in state.out_edges(node):
+            tree = forward_memlet_tree_with_nested_and_copies(state, out_edge)
+
+            while tree.parent is not None:
+                tree = tree.parent
+
+            print(print_tree(tree))
```
Comment (tbennun) [inline daceml/transformation/input_to_constant.py:204]:
```suggestion
```

File: daceml/transformation/reshape_elimination.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    while len(states) > 0:
+        state = states.pop()
+        expanded_something = False
+        for node in list(state.nodes()):  # Make sure we have a copy
+            if isinstance(node, nodes.NestedSDFG):
+                node.sdfg.expand_library_nodes()  # Call recursively
+            elif isinstance(node, nodes.LibraryNode) and not isinstance(
+                    node, donnx.ONNXReshape):
+                impl_name = node.expand(self, state)
+                print(
```
Comment (tbennun) [inline daceml/transformation/reshape_elimination.py:25]: add `if debugprint`

File: daceml/transformation/reshape_elimination.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+                _prod(desc.shape[i + 1:]) for i in range(len(desc.shape))
+            ]
+            return all(es == s
+                       for es, s in zip(expected_strides, desc.strides))
+
+        for desc in [
+                sdfg.arrays[in_memlet.data], sdfg.arrays[access_node.data]
+        ]:
+            if not is_desc_contiguous(desc):
+                return False
```
Comment (tbennun) [inline daceml/transformation/reshape_elimination.py:97]: Wouldn't all you want to check here is "if subset.coord_at(subset.min_element()) == view_subset.coord_at(view_subset.min_element()" (i.e., same pointer)?
@orausch

File: daceml/transformation/reshape_elimination.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    # pattern matching only checks that the type of the node matches,
+    _reshape_node = xf.PatternNode(donnx.ONNXReshape)
+    _access_node = xf.PatternNode(nodes.AccessNode)
+    _nsdfg = xf.PatternNode(nodes.NestedSDFG)
+
+    @staticmethod
+    def expressions():
+        return [
+            sdfg_utils.node_path_graph(ReshapeElimination._reshape_node,
```
Comment (tbennun) [inline daceml/transformation/reshape_elimination.py:49]: oh, now I'm confused. I thought it was nsdfg->reshape->access or access->reshape->nsdfg?

File: daceml/transformation/reshape_elimination.py [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+        # get the edges between the the access_node and the nsdfg_node
+        old_edges = [
+            e for e in state.out_edges(access_node) if e.dst == nsdfg_node
+        ]
+
+        for edge in old_edges:
+            state.add_edge(old_edge_in.src, old_edge_in.src_conn, edge.dst,
+                           edge.dst_conn, old_edge_in.data)
+            state.remove_edge(edge)
```
Comment (tbennun) [inline daceml/transformation/reshape_elimination.py:140]: looks a bit excessive here. There are only ever two edges and that's it.

File: daceml/util/utils.py [hunk 20]
```diff
@@ -104,6 +105,54 @@ def find_str_not_in_set(existing: typing.Set[str],
     return base_name + "_" + str(i)
 
 
+def vectorize_array_and_memlet(sdfg, array_name, type: dtypes.typeclass):
+    '''
+       Adjust the shape of a data container according to the vec width (only the last dimension).
```
Comment (tbennun) [inline daceml/util/utils.py:110]: Shouldn't -1 (the last dimension) just be a parameter of the function? i.e., "dim: int = -1"?

File: daceml/util/utils.py [hunk 21]
```diff
@@ -104,6 +105,54 @@ def find_str_not_in_set(existing: typing.Set[str],
...
+
+    # Search for all the memlets
+    for state in sdfg.nodes():
+        for edge in state.edges():
+            if edge.data.data == array_name:
+                # get the range
+                start, stop, skip = edge.data.subset.ranges[-1]
+
+                # Let's be conservative for the moment
+
```
Comment (tbennun) [inline daceml/util/utils.py:145]:
```suggestion
```

File: examples/plot_fpga_lenet.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
+"""
```
Comment (tbennun) [inline examples/plot_fpga_lenet.py:1]: what do you mean by the name "plot" in the filename

File: tests/transformation/test_input_to_constant.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+    donnx.ONNXGemm.default_implementation = "pure"
+
+    net = TestModule()
+    dace_net = DaceModule(net, dummy_inputs=(torch.rand(10, 5), ))
+
+    inp = torch.rand((10, 5))
+    #
+    sdfg: dace.SDFG = dace_net.sdfg
+    sdfg.expand_library_nodes()
+    sdfg.apply_transformations_repeated([InputToConstant], print_report=True)
```
Comment (tbennun) [inline tests/transformation/test_input_to_constant.py:32]: you can just assert that inputtoconstant happened X times (the return value of apply_transformations_repeated is the total number of applied transformations).

**Review by tbennun — CHANGES_REQUESTED** [review]

File: daceml/transformation/reshape_elimination.py (hunk 18 above)
Comment (orausch) [reply daceml/transformation/reshape_elimination.py:49]: This is elimination of an `ONNXReshape`, not an Access node`

File: examples/plot_fpga_lenet.py (hunk 22 above)
Comment (orausch) [reply examples/plot_fpga_lenet.py:1]: `plot_` is the default prefix used by sphinx gallery. We can change it if we want, suggestions welcome.

File: daceml/onnx/op_implementations/fpga_implementations.py [hunk 24]
```diff
@@ -0,0 +1,3090 @@
...
+from dace.symbolic import symstr
+
+from daceml.onnx.nodes.onnx_op import ONNXOp
+from daceml.onnx.forward_implementation_abc import ONNXForward
+import numpy as np
+import math
+
+from daceml.util.utils import in_desc_with_name, out_desc_with_name, in_edge_with_name
+from daceml.transformation import constant_folding
+from daceml.onnx.op_implementations.utils import op_implementation, program_for_node
```
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply daceml/onnx/op_implementations/fpga_implementations.py:17]: yes, I agree. But we will need to have Xilinx  and better vectorization support first (which may also simplify the implementation of some operator....or complicate it)

File: daceml/transformation/reshape_elimination.py (hunk 17 above)
Comment (orausch) [reply daceml/transformation/reshape_elimination.py:97]: I thought the Nested SDFG entry only handled contiguous reshapes (i.e. no changes in strides). But this was written before view nodes existed, not sure if we even need this whole transformation anymore. I'd propose to just leave it and update if we hit new cases.

File: daceml/util/utils.py (hunk 20 above)
Comment (TizianoDeMatteis) [reply daceml/util/utils.py:110]: Yeah, but didn't we say that vectorization can occur only on the last dimension?

File: daceml/transformation/reshape_elimination.py (hunk 17 above)
Comment (tbennun) [reply daceml/transformation/reshape_elimination.py:97]: A nested SDFG used to be a primitive version of View. You can use the data in any type or shape/strides there, no restriction.

Sure, let's keep it as is.

File: daceml/util/utils.py (hunk 20 above)
Comment (tbennun) [reply daceml/util/utils.py:110]: depending on the strides, but ok, we can assume C-contiguousness.

File: examples/plot_fpga_lenet.py (hunk 22 above)
Comment (tbennun) [reply examples/plot_fpga_lenet.py:1]: ok, what do you think about `demo_`?

**Comment by orausch** [comment]
@TizianoDeMatteis Seems like the main reason for the bad coverage is because reshape elimination is not tests

**Review by tbennun — APPROVED** [review]
